### ISSUE TITLE
`torch.ops.aten.index_put` returns different results on CUDA and CPU

### ISSUE THREAD

**Issue by hoshibara** [pytorch/pytorch#156173]
## Describtion
While debugging the cases related to flex attention learnable bias, I found some accuracy issues related to `test_head_specific_gate`.
It appears that the output of `torch.ops.aten.index_put` might be incorrect.
For example, the results are different between CPU and CUDA.

## Reproducer
```python
import torch

torch.manual_seed(42)
device = 'cuda'  # change to cpu

dim1_indices = torch.arange(4, dtype=torch.int64).view(1, 4, 1, 1)
indice = dim1_indices.expand((2, 4, 256, 256))
vals = torch.randn((2, 4, 256, 256), dtype=torch.float16)
grad = torch.zeros((4,), dtype=torch.float16)
accumulate = True

indice = indice.to(device)
vals = vals.to(device)
grad = grad.to(device)

print("[DEBUG] grad:\n", grad)
print("[DEBUG] indice:\n", indice)
print("[DEBUG] vals:\n", vals)
print("[DEBUG] accumulate:", accumulate)

result = torch.ops.aten.index_put(grad, [indice], vals, accumulate=accumulate)
print("[DEBUG] result:\n", result)


# on CPU:
# tensor([-373.0000, -121.8125,  -67.1875, -558.0000], dtype=torch.float16)
# on CUDA:
# tensor([-323.0000, -121.5000,  -70.1875, -545.5000], device='cuda:0', dtype=torch.float16)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.8.0.dev20250610+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-2ubuntu1~20.04) 11.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.10.16 | packaged by conda-forge | (main, Apr  8 2025, 20:53:32) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-214-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.4.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-PCIE-40GB
Nvidia driver version: 570.133.20
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      48 bits physical, 48 bits virtual
CPU(s):                             64
On-line CPU(s) list:                0-63
Thread(s) per core:                 1
Core(s) per socket:                 64
Socket(s):                          1
NUMA node(s):                       8
Vendor ID:                          AuthenticAMD
CPU family:                         25
Model:                              1
Model name:                         AMD EPYC 7713 64-Core Processor
Stepping:                           1
Frequency boost:                    enabled
CPU MHz:                            1498.952
CPU max MHz:                        2000.0000
CPU min MHz:                        1500.0000
BogoMIPS:                           4000.23
Virtualization:                     AMD-V
L1d cache:                          2 MiB
L1i cache:                          2 MiB
L2 cache:                           32 MiB
L3 cache:                           256 MiB
NUMA node0 CPU(s):                  0-7
NUMA node1 CPU(s):                  8-15
NUMA node2 CPU(s):                  16-23
NUMA node3 CPU(s):                  24-31
NUMA node4 CPU(s):                  32-39
NUMA node5 CPU(s):                  40-47
NUMA node6 CPU(s):                  48-55
NUMA node7 CPU(s):                  56-63
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==2.1.2
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.26.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] pytorch-triton==3.3.1+gitc8757738
[pip3] torch==2.8.0.dev20250610+cu126
[pip3] torchaudio==2.8.0.dev20250610+cu126
[pip3] torchvision==0.23.0.dev20250610+cu126
[pip3] triton==3.3.0
[conda] numpy                     2.1.2                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.6.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.6.80                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.6.77                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.6.77                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.5.1.17                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.0.4                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.7.77                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.1.2                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.4.2                 pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.7.1                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.26.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.85                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.6.77                  pypi_0    pypi
[conda] pytorch-triton            3.3.1+gitc8757738          pypi_0    pypi
[conda] torch                     2.8.0.dev20250610+cu126          pypi_0    pypi
[conda] torchaudio                2.8.0.dev20250610+cu126          pypi_0    pypi
[conda] torchvision               0.23.0.dev20250610+cu126          pypi_0    pypi
[conda] triton                    3.3.0                    pypi_0    pypi
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

**Comment by Flink-ddd** [comment]
Hi @hoshibara ,

Thank you for the excellent and detailed report. I've looked into this, and it seems to be a classic case of precision differences between CPU and parallel GPU operations.

The root cause of the discrepancy is the combination of two factors:

1.  Non-Associativity of Floating-Point Addition: As per the IEEE 754 standard, floating-point math is not associative. This means the order of operations matters, i.e., `(x + y) + z` may not yield the exact same result as `x + (y + z)`.
2.  Non-Deterministic Execution on GPU: Your `indice` tensor causes many values from `vals` to be accumulated into the same elements of the `grad` tensor. To handle these parallel writes without data loss, the GPU backend uses atomic operations (like `atomicAdd`). The execution order of these atomic operations across thousands of threads is non-deterministic.

The CPU computes the sum sequentially and deterministically, while the GPU computes it in a parallel, non-deterministic order. This difference in summation order leads to the different results you're observing. The effect is particularly pronounced with `torch.float16` due to its lower precision.

Besides accepting this inherent numerical difference, here are a couple of things you could try:

Use Higher Precision: Switching `dtype` to `torch.float32` will likely reduce the error margin significantly, as `float32` has much better precision for accumulation.

Enforce Determinism: For debugging or if strict reproducibility is required, you can try enabling deterministic algorithms:
    ```python
    torch.use_deterministic_algorithms(True)
    ```
    However, please be aware that this can impact performance and might not have a deterministic implementation for every operation on CUDA. This seems to be a fundamental trade-off between performance and bit-wise reproducibility in parallel computing.

**Comment by colesbury** [comment]
CUDA results look good, but the cpu results look off to me even for fp16. Slight variation with fp64 for expected values:

```python
import torch
import sys

seed = int(sys.argv[2]) if len(sys.argv) > 2 else 42
device =sys.argv[1] if len(sys.argv) > 1 else "cuda"
print(f"{device=}")

torch.manual_seed(seed)

dim1_indices = torch.arange(4, dtype=torch.int64).view(1, 4, 1, 1)
indice = dim1_indices.expand((2, 4, 256, 256))
vals = torch.randn((2, 4, 256, 256), dtype=torch.float16)
grad = torch.zeros((4,), dtype=torch.float16)
accumulate = True

indice = indice.to(device)
vals = vals.to(device)
grad = grad.to(device)

print("[DEBUG] grad:\n", grad)
print("[DEBUG] indice:\n", indice)
print("[DEBUG] vals:\n", vals)
print("[DEBUG] accumulate:", accumulate)

result = torch.ops.aten.index_put(grad, [indice], vals, accumulate=accumulate)
print("[DEBUG] result:\n", result)

indice2 = indice.to(device)
vals2 = vals.to(device, dtype=torch.float64)
grad2 = grad.to(device, dtype=torch.float64)
result2 = torch.ops.aten.index_put(grad2, [indice2], vals2, accumulate=accumulate)
print("[DEBUG] expected:\n", result2)
```

CPU:

```
[DEBUG] result:
 tensor([604.0000,  69.7500, 314.2500, 324.7500], dtype=torch.float16)
[DEBUG] expected:
 tensor([608.4469,  59.1547, 337.9794, 350.0752], dtype=torch.float64)
```

vs. CUDA

```
[DEBUG] result:
 tensor([608.5000,  59.1562, 338.0000, 350.0000], device='cuda:0',
       dtype=torch.float16)
[DEBUG] expected:
 tensor([608.4469,  59.1547, 337.9794, 350.0752], device='cuda:0',
       dtype=torch.float64)
```

**Comment by ngimel** [comment]
CUDA computes intermediate results for repeating indices in fp32, cpu does not. Not a bug, as there are no guarantee on the precision of intermediate accumulation. Also this has nothing to do with `torch.use_deterministic_algorithms(True)`, results are already deterministic.

**Comment by colesbury** [comment]
Do we have a reason for doing fp16 accumulation in fp16 on CPU (vs. fp32) or is it just how we always have done it?

**Comment by ngimel** [comment]
We have always done it, generic fp16 support on cpu where there's no hardware support is mostly an afterthought.

**Comment by liangan1** [comment]
cc @leslie-fang-intel @CaoE

**Comment by nobody81127** [comment]
I applied a floating-point error analysis algorithm to calculate the error bounds for the result in cuda, and I found that the result in cpu falls within this range. Therefore, I believe this is indeed a floating-point error, likely caused by differences in the order of operations. @hoshibara @colesbury
